core/src/apps/bitcoin/sign_tx/approvers.py

```diff
@@ -0,0 +1,206 @@
+        if not self.authorization.check_sign_tx_input(txi, self.coin):
+            raise wire.ProcessError("Unauthorized path")
+
+        await super().add_internal_input(txi, amount)
+
+    def add_change_output(self, txo: TxOutputType, script_pubkey: bytes) -> None:
+        super().add_change_output(txo, script_pubkey)
+        self._add_output(txo, script_pubkey)
+        self.our_weight.add_output(script_pubkey)
+        self.group_our_count += 1
+
+    async def add_external_output(
+        self, txo: TxOutputType, script_pubkey: bytes
+    ) -> None:
+        await super().add_external_output(txo, script_pubkey)
+        self._add_output(txo, script_pubkey)
+
+    async def approve_tx(self) -> None:
+        # Ensure that at least one of the user's outputs is in a group with an external output.
+        if not self.anonymity:
+            raise wire.ProcessError("No anonymity gain")
+
+        # The mining fee of the transaction as a whole.
+        mining_fee = self.total_in - self.total_out
+
+        # mining_fee > (coin.maxfee per byte * tx size)
+        if mining_fee > (self.coin.maxfee_kb / 1000) * (self.weight.get_total() / 4):
+            raise wire.ProcessError("Mining fee over threshold")
+
+        # The maximum mining fee that the user should be paying.
+        our_max_mining_fee = (
+            mining_fee * self.our_weight.get_total() / self.weight.get_total()
+        )
+
+        # The coordinator fee for the user's outputs.
+        our_coordinator_fee = self._get_coordinator_fee()
+
+        # Total fees that the user is paying.
+        our_fees = self.total_in - self.external_in - self.change_out
+
+        if our_fees > our_coordinator_fee + our_max_mining_fee:
+            raise wire.ProcessError("Total fee over threshold")
+
+        if self.tx.lock_time > 0:
+            raise wire.ProcessError("nLockTime not allowed in CoinJoin")
+
+        if not self.authorization.check_sign_tx(self.tx, our_fees):
+            raise wire.ProcessError("Fees exceed authorized limit")
+
+    # Coordinator fee calculation.
+
+    def _get_coordinator_fee(self) -> float:
+        # Add the coordinator fee for the last group of outputs.
+        self._new_group(0)
+
+        return (
+            self.coordinator_fee_base
+            * self.authorization.fee_per_anonymity
+            / pow(10, FEE_PER_ANONYMITY_DECIMALS + 2)
+        )
+
+    def _add_output(self, txo: TxOutputType, script_pubkey: bytes):
+        # Assumption: CoinJoin outputs are grouped by amount.
+        if self.group_amount != txo.amount:
+            self._new_group(txo.amount)
+
+        self.group_size += 1
+
+    def _new_group(self, amount: int):
+        # Add the base coordinator fee for the previous group of outputs.
+        if self.group_size > 1:
+            self.coordinator_fee_base += (
+                self.group_our_count * self.group_size * self.group_amount
+            )
+
+        # Check whether our outputs gained any anonymity.
```

Comment:
Yes, I don't think we can do any better.

Reply:
If our output is in a group of size 1, then it's a change output. If we have n outputs in a group of size n, then it could be a botched CoinJoin or it could just happen to be a group of our own change outputs, which just happen to have the same value. Hmm, unless we always have only one change output ...

Reply:
It seems to me in Wasabi you should be able to register multiple inputs into one CoinJoin, especially if they will each be using different output denominations, and in that case you want a separate change-output for each input.

Reply:
sooo do we gain anonymity in case some other party happens to have a change output of the same size as we do?

Reply:
Technically we do, but we are not aware of it or at least we didn't expect it and neither did the other party, so it's possible that it won't be treated as a CoinJoined output, e.g. because it will go to some non-anonymous change account path. This of course depends on the client.

I see this as a mere sanity check, that the transaction at least looks like a coinjoin giving us some anonymity.